File: LaboratoryWork/Semester_3/Laboratory_2/GoLang/hash/lruHash.go
package hash

import (
	"fmt"
	"math"
)

type CacheItem struct {
	key       int
	value     int
	timestamp int
	valid     bool
}

type LRUCache struct {
	capacity int
	timer    int
	cache    []CacheItem
}

func NewLRUCache(cap int) *LRUCache {
	cache := make([]CacheItem, cap)
	return &LRUCache{
		capacity: cap,
		timer:    0,
		cache:    cache,
	}
}

func (l *LRUCache) findIndex(key int) int {
	for i := range l.cache {
		if l.cache[i].valid && l.cache[i].key == key {
			return i
		}
	}
	return -1
}

func (l *LRUCache) findLRUIndex() int {
	oldestTime := math.MaxInt
	oldestIndex := -1
	for i := range l.cache {
		if l.cache[i].valid && l.cache[i].timestamp < oldestTime {
			oldestTime = l.cache[i].timestamp
			oldestIndex = i
		}
	}
	return oldestIndex
}

func (l *LRUCache) findFreeSlot() int {
	for i := range l.cache {
		if !l.cache[i].valid {
			return i
		}
	}
	return -1
}

func (l *LRUCache) Get(key int) int {
	index := l.findIndex(key)
	if index == -1 {
		return -1
	}
	l.timer++
	l.cache[index].timestamp = l.timer
	return l.cache[index].value
}

func (l *LRUCache) Insert(key int, value int) {
	index := l.findIndex(key)
	if index != -1 {
		l.cache[index].value = value
		l.timer++
		l.cache[index].timestamp = l.timer
		return
	}

	freeSlot := l.findFreeSlot()
	l.timer++
	if freeSlot != -1 {
		l.cache[freeSlot] = CacheItem{
			key:       key,
			value:     value,
			timestamp: l.timer,
			valid:     true,
		}
	} else {
		lruIndex := l.findLRUIndex()
		if lruIndex != -1 {
			l.cache[lruIndex] = CacheItem{
				key:       key,
				value:     value,
				timestamp: l.timer,
				valid:     true,
			}
		}
	}
}

func (l *LRUCache) Remove(key int) bool {
	index := l.findIndex(key)
	if index == -1 {
		return false
	}
	l.cache[index].valid = false
	return true
}

func (l *LRUCache) Display() {
	fmt.Printf("LRUCache (capacity: %d, size: %d):\n", l.capacity, l.GetSize())
	for _, item := range l.cache {
		if item.valid {
			fmt.Printf("  [%d -> %d, ts: %d]\n", item.key, item.value, item.timestamp)
		}
	}
}

func (l *LRUCache) GetSize() int {
	count := 0
	for _, item := range l.cache {
		if item.valid {
			count++
		}
	}
	return count
}

func (l *LRUCache) Empty() bool {
	return l.GetSize() == 0
}

func (l *LRUCache) Keys() []int {
	keys := []int{}
	for _, item := range l.cache {
		if item.valid {
			keys = append(keys, item.key)
		}
	}
	return keys
}

func (l *LRUCache) Values() []int {
	values := []int{}
	for _, item := range l.cache {
		if item.valid {
			values = append(values, item.value)
		}
	}
	return values
}

func (l *LRUCache) Clear() {
	for i := range l.cache {
		l.cache[i].valid = false
	}
	l.timer = 0
}

File: LaboratoryWork/Semester_3/Laboratory_2/C++/src/tasks/LRUHash.cpp
#include "../../include/LRUHash.hpp"

LRUCache::CacheItem::CacheItem() : key(-1), value(-1), timestamp(-1), valid(false) {}

LRUCache::LRUCache(int cap) : capacity(cap), timer(0), cache(cap) {}

int LRUCache::findIndex(int key) const {
    for (size_t i = 0; i < cache.size(); i++) {
        if (cache[i].valid && cache[i].key == key)
            return i;
    }
    return -1;
}

int LRUCache::findLRUIndex() const {
    int oldestTime = numeric_limits<int>::max();
    int oldestIndex = -1;

    for (size_t i = 0; i < cache.size(); i++) {
        if (cache[i].valid && cache[i].timestamp < oldestTime) {
            oldestTime = cache[i].timestamp;
            oldestIndex = i;
        }
    }
    return oldestIndex;
}

int LRUCache::findFreeSlot() const {
    for (size_t i = 0; i < cache.size(); i++) {
        if (!cache[i].valid)
            return i;
    }
    return -1;
}

int LRUCache::get(int key) {
    int index = findIndex(key);
    if (index == -1) return -1;

    cache[index].timestamp = ++timer;
    return cache[index].value;
}

void LRUCache::insert(int key, int value) {
    int index = findIndex(key);

    if (index != -1) {
        cache[index].value = value;
        cache[index].timestamp = ++timer;
        return;
    }

    int freeSlot = findFreeSlot();
    if (freeSlot != -1) {
        cache[freeSlot].key = key;
        cache[freeSlot].value = value;
        cache[freeSlot].timestamp = ++timer;
        cache[freeSlot].valid = true;
    } else {
        int lruIndex = findLRUIndex();
        if (lruIndex != -1) {
            cache[lruIndex].key = key;
            cache[lruIndex].value = value;
            cache[lruIndex].timestamp = ++timer;
            cache[lruIndex].valid = true;
        }
    }
}

bool LRUCache::remove(int key) {
    int index = findIndex(key);
    if (index == -1) return false;

    cache[index].valid = false;
    return true;
}

void LRUCache::display() const {
    cout << "LRUCache (capacity: " << capacity << ", size: " << getSize() << "):" << endl;
    for (size_t i = 0; i < cache.size(); i++) {
        if (cache[i].valid) {
            cout << "  [" << cache[i].key << " -> " << cache[i].value
                 << ", ts: " << cache[i].timestamp << "]" << endl;
        }
    }
}

int LRUCache::getSize() const {
    int count = 0;
    for (size_t i = 0; i < cache.size(); i++)
        if (cache[i].valid) count++;
    return count;
}

bool LRUCache::empty() const {
    return getSize() == 0;
}

DynamicArray<int> LRUCache::keys() const {
    DynamicArray<int> result;
    for (size_t i = 0; i < cache.size(); i++) {
        if (cache[i].valid)
            result.push_back(cache[i].key);
    }
    return result;
}

DynamicArray<int> LRUCache::values() const {
    DynamicArray<int> result;
    for (size_t i = 0; i < cache.size(); i++) {
        if (cache[i].valid)
            result.push_back(cache[i].value);
    }
    return result;
}

LRUCache::~LRUCache() {
    for (size_t i = 0; i < cache.size(); i++)
        cache[i].valid = false;
    timer = 0;
}